### ISSUE TITLE
nonselection_glyph seems to be drawn after selection_glyph

### ISSUE THREAD

**Issue by birdsarah** [bokeh/bokeh#2116]
Using patches to draw a map and want to do selection as it works on the map here: http://africawatermap.aptivate.org/

This doesn't seem to be possible as the non-selected glyphs are being drawn over the selected glyph. Couple of screenshots to highlight this. 

----

selected and non-selected elements are full alpha, but selected has thick line
![1](https://cloud.githubusercontent.com/assets/1796208/6887071/51bb92a0-d608-11e4-86a7-0685a2f2bc99.png)

non-selected element keeps line alpha but has full alpha set to 0
![2](https://cloud.githubusercontent.com/assets/1796208/6887072/51bd40f0-d608-11e4-8298-78e4c5e57cd1.png)

non-selected element has line_alpha of 0 and fill_alpha of 0.2
![3](https://cloud.githubusercontent.com/assets/1796208/6887070/51bb37ec-d608-11e4-8933-7e630ec9acfd.png)

non-selected element has line_alpha of 0 and fill_alpha of 1
![4](https://cloud.githubusercontent.com/assets/1796208/6887084/b0e5554a-d608-11e4-931f-bf5821817c0f.png)

**Comment by bryevdv** [comment]
@birdsarah is it possible you are drawing the patches twice somehow? I don't understand this because the main render method splits all the indices into two groups, the selected and non-selected, and calls individual glyph render functions separately for each (disjoin) set of indices here:

https://github.com/bokeh/bokeh/blob/master/bokehjs/src/coffee/renderer/glyph/glyph_renderer.coffee#L117

It's certainly possible there is a real problem in BokehJS I would just be surprised not to see if affect every glyph in the same way.

**Comment by birdsarah** [comment]
Hi @bryevdv I'm not understanding your question, but you have hit on the problem

My understanding of the code you linked to:
````
do_render(ctx, selected, @selection_glyph)
do_render(ctx, nonselected, @nonselection_glyph)
````

is that:
* It's rendering the non-selected elements second.
* This means that in the case where they overlap, the non-selected ones are "on top"

Having tried changing that order locally. That is indeed the source of my error:

With the current code:
![selected_rendered_before](https://cloud.githubusercontent.com/assets/1796208/6904910/0a0428ce-d6d9-11e4-8e61-cc7b435dca24.png)

With the order of do_render switched:
![selected_rendered_after](https://cloud.githubusercontent.com/assets/1796208/6904932/19675d22-d6d9-11e4-8eb0-2336742feb02.png)

In case it's useful, here's the relevant piece of code:

````python
    borders = Patches(
        xs='xs', ys='ys',
        fill_color='color_for_active_year', fill_alpha=1,
        line_color="#FFFFFF", line_width=1,
    )
    selected_borders = Patches(
        xs='xs', ys='ys',
        fill_color='color_for_active_year', fill_alpha=1,
        line_color=ORANGE, line_width=5,
    )

    plot.add_glyph(source, borders, selection_glyph=selected_borders, nonselection_glyph=borders)
````

**Comment by bryevdv** [comment]
Oh, I see. I misunderstood to mean that an entire non selected glyph was obscuring a selected one, not just the borders. That is a simple fix, but it should go in after our big BokehJS build PR which will hopefully be merged tomorrow.

**Comment by bryevdv** [comment]
This was changed.

**Comment by damianavila** [comment]
> This was changed.

Can you link where was changed? Thanks.

**Comment by bryevdv** [comment]
fixed in #2143
